tests/test_recipe_yaml/slepc_weak_strong.yaml

```diff
@@ -0,0 +1,42 @@
+schema_version: 1
+
+context:
+  version: 3.19.1
+  sha256: 280737e9ef762d7f0079ad3ad29913215c799ebf124651c723c1972f71fbc0db
+  build: 0
+
+package:
+  name: slepc
+  version: ${{ version }}
+
+source:
+  url: http://slepc.upv.es/download/distrib/slepc-${{ version }}.tar.gz
+  sha256: ${{ sha256 }}
+
+build:
+  number: ${{ build }}
+  string: real_h${{ PKG_HASH }}_${{ build }}
+  skip:
+    - win
+
+requirements:
+  run:
+    - petsc
+    - suitesparse
+  run_exports:
+    strong:
+      - ${{ pin_subpackage('slepc', upper_bound='x') }}
+    weak:
+      - ${{ pin_subpackage('slepc', upper_bound='x.x') }}
```

Comment:
The build strings and comments are important for this test as they catch an important edge case.

Reply:
I've re-added the comments. However, I cannot add the build strings.
This combination of `pin_subpackage` and build strings is not supported by rattler-build

Reply:
@beckermr - the way `pin_subpackage` and `pin_compatible` work in rattler-build it doesn't really allow any extra characters on the `pin_subpackage` line. We basically lazy evaluate the function at a later time in order to not have to do the crazy recursive eval of conda-build (which makes it slow). 

It might actually be something to test in `rattler-build`. I see two options to cover this case:

- we could add a new keyword argument to the function (`build: "foo_*"`)
- we could extend the parsing in a way that we parse the rest of the string as the build argument

I think the first of the two options would be much easier to implement. 

Maybe @h-vetinari has opinions because he triggered some larger re-shuffeling of the pin-function arguments :)

Reply:
For more context (@h-vetinari): the current requirement looks something like:

```yaml
- ${{ pin_subpackage(foo, max_pin="x.x") real_*  # some comment
```

Reply:
Ahhhh I see. We'll need something in the v2 spec for this as it is a use case we care about. Agreed that a `build` keyword is probably best.

Reply:
Yeah, I think a new kwarg sounds like a good idea! The previous reshuffling was about adding expressiveness (and reducing confusion because things didn't work as might be expected with the possibility to ask for contradictory things), so being able to to handle a case like [slepc](https://github.com/conda-forge/slepc-feedstock/blob/fdc04765da95151ad337532c24d72dc46932ff81/recipe/meta.yaml#L41-L44) (pin on additional constraints, but not as tight as `exact=True`) without workarounds sounds like a good addition to me.

How would we write this? 
```
  run_exports:
    - if: cuda_compiler_version == "None"
      then: ${{ pin_subpackage('slepc', upper_bound='x.x', build=scalar ~ '_*') }}
      else: ${{ pin_subpackage('slepc', upper_bound='x.x', build='cuda' ~ cuda_major ~ '_*') }}
```
Both `scalar` & `cuda_major` are variables; one from CBC, one derived from `cuda_compiler_version`. I'm assuming they are in context for when the function call happens? Or can we define a variable that corresponds to the build string (depending on the various combinations) and then just do:
```
  run_exports:
    - ${{ pin_subpackage('slepc', upper_bound='x.x', build=some_variable) }}
```

I think repeated application of `- if X: then: Y` makes the recipe pretty unreadable, so I'd definitely want to avoid a quadruple-if just for that run-export. Sidenote: are multi-line `then:` & `else:` clauses possible? I'm assuming they pytorch example here is a mechanical translation, because 
```
        - if: "cuda_compiler_version != \"None\""
          then: cudnn
        - if: "cuda_compiler_version != \"None\""
          then: nccl
        - if: "cuda_compiler_version != \"None\""
          then: magma
```
should really be
```
        - if: cuda_compiler_version != "None"
          then: cudnn
                nccl      # how do I write this?
                magma
```

Reply:
Yes, multi-line if-else – as you've typed – is possible. `conda-recipe-manager` doesn't do it nicely (for humans) yet.

Reply:
Personally, I don't mind the discussion here, but I wonder if we should open a new issue for this at https://github.com/prefix-dev/rattler-build so we can find it again later on

Reply:
I just created an issue: https://github.com/prefix-dev/rattler-build/issues/1012